src/miral/reloading_config_file.cpp

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright © Canonical Ltd.
+ *
+ * This program is free software: you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 or 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+#include <miral/reloading_config_file.h>
+
+#include <miral/runner.h>
+
+#include <mir/log.h>
+
+#include <boost/throw_exception.hpp>
+
+#include <sys/inotify.h>
+#include <unistd.h>
+
+#include <fstream>
+#include <optional>
+#include <string>
+#include <vector>
+
+using namespace std::filesystem;
+
+namespace
+{
+auto config_directory(path const& file) -> std::optional<path>
+{
+    if (file.has_parent_path())
+    {
+        return file.parent_path();
+    }
+    else if (auto config_home = getenv("XDG_CONFIG_HOME"))
+    {
+        return config_home;
+    }
+    else if (auto home = getenv("HOME"))
+    {
+        return  path(home) / ".config";
+    }
+    else
+    {
+        return std::nullopt;
+    }
+}
+
+auto watch_fd(mir::Fd const& inotify_fd, std::optional<path> const& path) -> std::optional<mir::Fd>
+{
+    if (!path.has_value())
+        return std::nullopt;
+
+    if (inotify_fd < 0)
+        BOOST_THROW_EXCEPTION((std::system_error{errno, std::system_category(), "Failed to initialize inotify_fd"}));
+
+    return mir::Fd{inotify_add_watch(inotify_fd, path.value().c_str(), IN_CLOSE_WRITE | IN_CREATE | IN_MOVED_TO)};
+}
+}
+
+class miral::ReloadingConfigFile::Self
+{
+public:
+    Self(MirRunner& runner, path file, Loader load_config);
+
+private:
+    mir::Fd const inotify_fd;
+    Loader const load_config;
+    std::filesystem::path const filename;
+    std::optional<std::filesystem::path> const directory;
+    std::optional<mir::Fd> const directory_watch_fd;
+
+    void register_handler(MirRunner& runner);
+    std::unique_ptr<miral::FdHandle> fd_handle;
+};
+
+miral::ReloadingConfigFile::Self::Self(MirRunner& runner, path file, Loader load_config) :
+    inotify_fd{inotify_init()},
```

Comment:
We probably want to be using `inotify_init1(IN_CLOEXEC)` here, no? Make sure the watch isn't leaked to child processes.